### ISSUE TITLE
Saving an empty table (heads only) as JSONL produces an empty file

### ISSUE THREAD

**Issue by daviewales** [saulpw/visidata#1726]
**Small description**

Saving an empty table with headers but no data as JSONL produces an empty file.

**Expected result**

A JSONL file including the headers.
(Compare to the output when saving a TSV)

**Actual result with screenshot**
If you get an unexpected error, please include the full stack trace that you get with `Ctrl-E`.

**Steps to reproduce with sample data and a .vd**
[test.vdj.zip](https://github.com/saulpw/visidata/files/10670604/test.vdj.zip)

Create a table directly in Visidata and save it as `.jsonl`.
The output file is empty.

**Additional context**
Please include the version of VisiData.
v2.11

**Comment by daviewales** [comment]
I'm guessing we need to grab the headers and save them first here?

https://github.com/saulpw/visidata/blob/472122901717ec912659ea09e6d43115a731ea07/visidata/loaders/json.py#L122

**Comment by saulpw** [comment]
Hm, I'm not sure understand what you need.  Each line in a JSONL file is a row, so without any rows, what should the actual file contain?

**Comment by daviewales** [comment]
Oh interesting! I was looking at the [JSONL example page](https://jsonlines.org/examples/), and assumed that Visidata's tabular data would be stored as per the JSONL example. e.g:

``` jsonl
["Name", "Session", "Score", "Completed"]
["Gilbert", "2013", 24, true]
["Alexa", "2013", 29, true]
["May", "2012B", 14, false]
["Deloise", "2012A", 19, true] 
```

It would make sense to be able to save the headers in this case.

But looking at the output from Visidata, I see that it uses rows of dicts, rather than lists.
Interestingly if I try to read the example JSONL data above with Visidata it just comes in as a single unnamed column:

![image](https://user-images.githubusercontent.com/2499066/217149576-aa13aa84-9a6b-47b7-a231-7e25cc0e89e7.png)

Is there a reason why Visidata uses the rows of dicts format rather than the rows of lists format?
Would it make sense to detect and support both styles for reading and writing?

**Comment by saulpw** [comment]
Huh!  I guess I never looked at the examples page, or I completely forgot about it.  For my money, the rows of dicts format is what confers all the advantages of the JSONL format, like being able to add a field to a single row without having to add empty values to all other rows.  It would be reasonable to add support for loading the rows of lists format; it wouldn't be too hard to add a saver too.  Is this rows of lists format in active use?   How are you using JSONL in practice?

**Comment by daviewales** [comment]
Hmm... I guess which is better depends on whether your dataset has lots of nulls or not. If you have no nulls, rows of lists will be more compact, while still retaining the JSON type information.

I have not used JSONL before today.
I was trialling it as a better defined alternative to TSV or CSV for a very small handwritten table. I created the headers in Visidata, saved the table and exited, and was very surprised when the file was empty!

Perhaps if Visidata gains the ability to load rows of lists JSONL, it could default to saving JSONL as rows of lists if there are only headers defined?

**Comment by saulpw** [comment]
> Perhaps if Visidata gains the ability to load rows of lists JSONL, it could default to saving JSONL as rows of lists if there are only headers defined?

Hm, this seems maybe a bit too 'magic'--the actual format shouldn't change based on the number of rows.  There could be an option, `json_save_arrays` or something, which would cause it to save as arrays instead of dicts.  And this option could be set automatically on a JSON sheet, if the json loader finds itself loading a list of arrays rather than dicts.

**Comment by daviewales** [comment]
I see what you mean. It's more consistent to keep the format the same. But it is very surprising to find an empty file when saving a file with headers only!

Would it make sense to print a warning when a completely empty file is saved? Perhaps noting the option to save in a different format to keep the headers?

**Comment by saulpw** [comment]
I can see why that would be surprising!  It's a good idea to add a warning.  How about `"No rows saved."`?  and what if we expanded the warning system to put an extended warning message in the sidebar?

```
A JSONL file without any rows is an empty file; is that what you want?
Try `options.json_save_arrays=True` to save as JSONL array format, rather than JSONL dict format,
if you truly want the headers only.
```

**Comment by daviewales** [comment]
That looks perfect!

(There's a sidebar?!)

**Comment by daviewales** [comment]
I'm happy to take a stab at adding this. Would you agree to the following changes?

- [x] Add warning when saving `jsonl` file with no rows
- [x] Add JSONL arrays loader
- [x] Add JSONL arrays saver
- [x] Add option to save JSONL as arrays
- [x] Add extended warning which hints to use JSONL arrays format to preserve headers

**Comment by saulpw** [comment]
That would be great, @daviewales!

**Comment by daviewales** [comment]
@saulpw, do we want to create an empty file and give a warning "jsonl output file is empty", or do we want to error and say "can't save jsonl file without data"?

**Comment by daviewales** [comment]
And do we want to warn if any columns are skipped, or only if the file will be empty?

**Comment by saulpw** [comment]
> do we want to create an empty file and give a warning "jsonl output file is empty", or do we want to error and say "can't save jsonl file without data"?

This is a great question.  I can imagine a replay script that selects all matching entries and then saves them off before executing another unrelated command, and a `fail()` in that case would abort the replay.  So I think we want `warning()` instead of `fail()` (`error()` should be used for code errors or external problems, not commands that have failed for 'valid' reasons).

**Comment by saulpw** [comment]
> And do we want to warn if any columns are skipped, or only if the file will be empty?

Hm, I would think only if the file is empty.  In what case would we skip columns?

**Comment by daviewales** [comment]
If you save this file, it will only contain column A. Column B will be skipped:

![image](https://user-images.githubusercontent.com/2499066/217966673-1687f4a2-ba22-4ff6-8bea-372101280289.png)

**Comment by daviewales** [comment]
If the file will be completely empty, should we skip creating the file and show a warning, or should we create the empty file and show a warning?

**Comment by saulpw** [comment]
I'd say we should create the empty file and show a warning.  Things shouldn't fail just because there isn't any data; and if something is using this as part of the pipeline, if there isn't any data for some reason, the rest of the pipeline should be able to continue with no data (and if it can't deal with an empty file, that's its problem).

**Comment by saulpw** [comment]
> If you save this file, it will only contain column A. Column B will be skipped:

Ah, I see.  I think we shouldn't warn in this case, because it would take too much time and code to track whether a column ever has non-null values.  And we definitely shouldn't warn about every null value.

**Comment by daviewales** [comment]
I've got something working which detects the JSONL headers array format automatically using the heuristic 'If the first row is a list containing string values only then it's a JSONL arrays header'.

Is this an OK approach? Or would you prefer the user to explicitly choose the arrays format?

**Comment by saulpw** [comment]
For now, let's make it explicit.  I've been thinking about how to detect file formats and options more generally and this would fit in with that.

**Comment by daviewales** [comment]
Closing this as it's addressed by #1730
